numpy/_core/src/umath/ufunc_object.c

```diff
@@ -4564,9 +4586,24 @@ ufunc_generic_fastcall(PyUFuncObject *ufunc,
             Py_DECREF(operands[i]);
         }
     }
+
+    /*
+     * all_inputs_were_scalars is used to decide if 0-D results should be
+     * unpacked to a scalar.  But, `np.matmul(vector, vector)` should do this.
+     * So for now we assume that if gufuncs return a 0-D result, we should
+     * unpack.  (unless keepdims=True, since that behave like a normal ufunc)
+     * So we pretend inputs were scalars...
+     *
+     * TODO: We may need a way to customize that at some point.
```

Comment:
For gufunc, whether an argument is scalar would seem to have the logical condition that there are no outer dimensions being iterated over. The effect would be the same as what you have here, but perhaps easier to describe?

Reply:
Yeah, maybe that is a better way to describe "no outer dimensions in the result".

Unfortunately, that is also a nice way to explain why we don't like this behavior :).  I.e. the whole point of this is that it means the number of outer dimensions (zero or more) has never an impact on the result type!
Unfortunately, unlike `arr.sum()`, `vecdot(x, y)` doesn't have a way to signal that we should return a scalar (by default).
Maybe `out=...` has to be the one way unfortunately... Or maybe `vecdot` should return arrays.  Also, we could (and probably should!) make it so that if `axes=` is passed we never return a scalar (mirroring `arr.sum(0)`).

Reply:
> Instead, maybe we should have an equivalent of out=... that makes explicit one does want a scalar if possible, say, out=() (or out=((), ()) for multiple outputs)

Maybe, I omitted multiple outputs for `...` and I think it is better here as well.  If we are going to go a step further here, I am wondering if even a `return_scalar=True/False/None` may not be better, though (and remove `out=...` again).

Maybe the main thing is if `vec @ vec` can change behavior.  Because if we are willing to gamble on that, it may also be OK to just ask users to do `np.matmul(vec, vec)[()]` for starters.  Or a more obvious `arr.get_element()`